README.md
<h1 align="center">Home Dashboard Weather Manager </h1>
<p align="center">
    <img src="https://travis-ci.com/iamtomhewitt/home-dashboard-weather-manager.svg"/>
    <img src="https://heroku-badge.herokuapp.com/?app=home-dashboard-weather-manager&style=round&svg=1"/>
</p>
<p align="center">
    ☀ A piggy back API of <a href="https://darksky.net/dev">Darksky</a> to get the weather for my home dashboard.
</p>

## Pipeline
* `Travis` tests the repo using `npm test`, which runs `mocha 'tests/**/*.js' --exit`
* Once the tests pass, `Heroku` deploys the app.
* When the app is deployed, you can make requests to it.

## Endpoints

### `/ (GET)`
* The root endpoint, returning information about the app.

#### Responses
* `200` success
```json
{
    "status": "🌞 SERVER OK",
    "version": "1.0.0",
    "endpoints": [
        {
            "path": "/weather",
            "methods": [
                "GET"
            ]
        },
        {
            "path": "/",
            "methods": [
                "GET"
            ]
        }
    ]
}
```

### `/weather (GET)`
* Gets the weather using latitude and longitude codes.
* Query parameters:
	* `latitude=<latitude>`
	* `longitude=<longitude>`
	* `apiKey=<key>`

#### Responses
* `200` success
```json
{
    "current": {
        "summary": "A summary of the weather",
        "icon": "sunshine",
        "temperature": 15
    },
    "week": [
        {
            "day": "<day>",
            "icon": "cloudy",
            "temperatureLow": 6,
            "temperatureHigh": 10,
        },
        {
            "day": "<day>",
            "icon": "rain",
            "temperatureLow": 4,
            "temperatureHigh": 9,
        },
        ...
    ]
}
```
* `400` bad request if there are missing parameters
```json
{
    "status": 400,
    "message": "There are missing query parameters"
}
```
* `401` unauthorised if api key is incorrect
```json
{
    "status": 401,
    "message": "API key is incorrect"
}
```
* `500` if the weather could not be retrieved
```json
{
    "status": 500,
    "message": "Could not get weather: <error message>"
}
```


CHANGELOG.md
## 1.0.3 (06/09/2020) 


### Issues in this release:

* [#3](https://github.com/iamtomhewitt/home-dashboard-weather-manager/issues/3) - Use github actions



## 1.0.2 (14/07/2020) 


There are no issues in this release.

tests/tests.js
const request = require('supertest');
const assert = require('assert');
require('dotenv').config();

describe('/ tests', () => {
    let server;

    before(() => {
        server = require('../app').listen(3002);
    });

    after(() => {
        server.close();
    });

    it('/ gives 200', (done) => {
        request(server)
            .get('/')
            .expect(200)
            .end((err, response) => {
                if (err) {
                    return done(err);
                }
                const { version } = require('../package.json');
                const status = '🌞 SERVER OK';

                assert.equal(response.body.version, version);
                assert.equal(response.body.status, status);

                return done();
            });
    });
});

describe('/weather tests', () => {
    let server;

    before(() => {
        server = require('../app').listen(3002);
    });

    after(() => {
        server.close();
    });

    it('/weather gives 200', (done) => {
        request(server)
            .get(`/weather?latitude=37.8267&longitude=-122.4233&apiKey=${process.env.API_KEY}`)
            .expect(200)
            .end((err, response) => {
                if (err) {
                    return done(err);
                }

                assert.notEqual(response.body.current, null);
                assert.notEqual(response.body.current.summary, null);
                assert.notEqual(response.body.current.icon, null);
                assert.notEqual(response.body.current.temperature, null);

                assert.notEqual(response.body.week, null);
                assert.notEqual(response.body.week[0].day, null);
                assert.notEqual(response.body.week[0].icon, null);
                assert.notEqual(response.body.week[0].temperatureLow, null);
                assert.notEqual(response.body.week[0].temperatureHigh, null);

                return done();
            });
    });

    it('/weather gives 400 if query parameters are missing', (done) => {
        request(server)
            .get('/weather')
            .expect(400)
            .end((err, response) => {
                if (err) {
                    return done(err);
                }

                assert.equal(response.body.status, 400);
                assert.equal(response.body.message, 'There are missing query parameters');

                return done();
            });
    });

    it('/weather gives 401 if API key is incorrect', (done) => {
        request(server)
            .get('/weather?latitude=10&longitude=5&apiKey=incorrect')
            .expect(401)
            .end((err, response) => {
                if (err) {
                    return done(err);
                }

                assert.equal(response.body.status, 401);
                assert.equal(response.body.message, 'API key is incorrect');

                return done();
            });
    });
});


app.js
const express = require('express');
const bodyParser = require('body-parser');
const listEndpoints = require('express-list-endpoints');
const request = require('request');
const { version } = require('./package.json');

const app = express();

const success = 200;
const badRequest = 400;
const unauthorised = 401;
const serverError = 500;

function response(status, message) {
    return {
        status,
        message,
    };
}

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));

app.get('/', (req, res) => {
    res.status(success).send({ status: '🌞 SERVER OK', version, endpoints: listEndpoints(app) });
});

app.get('/weather', (req, res) => {
    const { latitude } = req.query;
    const { longitude } = req.query;
    const { apiKey } = req.query;

    if (!latitude || !longitude || !apiKey) {
        res.status(badRequest).send(response(badRequest, 'There are missing query parameters'));
        return;
    }

    const url = `https://api.darksky.net/forecast/${apiKey}/${latitude},${longitude}?units=uk`;

    request(url, (err, resp) => {
        if (err) {
            res.status(serverError).send(response(serverError, err.message));
            return;
        }

        if (resp.statusCode === 403) {
            res.status(unauthorised).send(response(unauthorised, 'API key is incorrect'));
            return;
        }

        if (resp.statusCode === 200) {
            const days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
            const darkskyResponse = JSON.parse(resp.body);
            const current = {
                summary: darkskyResponse.currently.summary,
                icon: darkskyResponse.currently.icon,
                temperature: darkskyResponse.currently.temperature,
            };
            const week = [];

            for (let i = 1; i < darkskyResponse.daily.data.length; i += 1) {
                const date = new Date(darkskyResponse.daily.data[i].time * 1000);
                const day = days[date.getDay()];
                const icon = darkskyResponse.daily.data[i].icon;
                const temperatureLow = darkskyResponse.daily.data[i].temperatureLow;
                const temperatureHigh = darkskyResponse.daily.data[i].temperatureHigh;

                week.push({
                    day,
                    icon,
                    temperatureLow,
                    temperatureHigh,
                });
            }

            res.status(success).send({ current, week });
        } else {
            res.status(serverError).send(response(serverError, `Could not get weather: ${resp}`));
        }
    });
});

const port = 3001;
app.listen(process.env.PORT || port, () => { });

module.exports = app;
